projenrc/TypecheckTests.ts

```diff
@@ -0,0 +1,58 @@
+import { Component, JsonFile } from 'projen';
+import type { TypescriptConfig } from 'projen/lib/javascript';
+import type { TypeScriptProject } from 'projen/lib/typescript';
+
+/**
+ * Enable type checking for the test of the given project
+ *
+ * This creates a `tsconfig.json` in the test directory with a `noEmit: true`
+ * directive that inherits from `tsconfig.dev.json`, and adds a command job to
+ * the `compile` step.
+ *
+ * This necessary because during refactorings it's very easy for type errors
+ * to creep into tests, and the reporting of these errors only happens during
+ * testing running otherwise which is annoying.
+ *
+ * # This looks crazy
+ *
+ * The needs to be called `tsconfig.json` otherwise VSCode won't load it.
+ *
+ * That means the file must live in `test` otherwise it will conflict with other
+ * `tsconfig`s that I don't want to even think about having to rename. The
+ * easiest way to add a `test/tsconfig.json` without disturbing too much is to
+ * extend the parent `tsconfig.dev.json`, but we have to change the rootDir
+ * and also we have to copy over references because those are not inherited.
+ */
+export class TypecheckTests extends Component {
+  constructor(repo: TypeScriptProject) {
+    super(repo);
+
+    new JsonFile(repo, 'test/tsconfig.json', {
+      obj: {
+        extends: '../tsconfig.dev.json',
```

Comment:
Will this cause a recompilation of non test files though? 

https://github.com/aws/aws-cdk-cli/blob/25db152d769ab547fe71259fb2d9e020cbb0a46b/packages/aws-cdk/tsconfig.dev.json#L34-L35

If so - lets override `include` to only point to test files?

Reply:
It will cause a recompilation of test files. If we change `include` to non-test files, compilation will fail because we typically `import` files from `lib` and they need to part of the "compiled source set".

The 100% correct way of doing is this to put an independent `tsconfig.json` in every directory, and use project references to link them up. So:

```
packages/
    lib1/
        src/
	    tsconfig.json
	test/
	    tsconfig.json  { references: ../src }
    lib2/
        src/
	    tsconfig.json  { references: ../../lib1/src }
	test/
	    tsconfig.json  { references: ../src, ../../lib1/src }
```

I don't want to spend a week getting that right.